### ISSUE TITLE
[BUG] Dartagnan fails with error `Loop entry label 12:$bb3: has multiple back edges`

### ISSUE THREAD

**Issue by mProjectsCode** [hernanponcedeleon/Dat3M#465]
With some specific input code we get this error when verifying with dartagnan.

Verification failed with exception:
```
com.dat3m.dartagnan.exception.MalformedProgramException: Loop entry label 12:$bb3: has multiple back edges
```

**Steps to reproduce**

Folder Structure
```
<Folder>
|_main.c
|_userCode.h
|_userCode.c
```

`main.c`
```c
#include <stdio.h>
#include <pthread.h>

#include "userCode.h"

//Test function removes every even number
void *test1(){
    for(int i = 2; i < 1000; i++) {
        if(insert(i) == 0){
            printf("Number already in Set \n");
        } else {
            printf("thread1: inserted %d \n", i);
        }
        if(i % 2 == 0) {
            delete(i);
            printf("thread1: deleted %d \n", i);
        }
    }
}

//test function removes every third number
void *test2(){
    for(int i = 2; i < 1000; i++) {
        insert(i);
        printf("thread2: inserted %d \n", i);
        if(i % 3 == 0) {
            delete(i);
            printf("thread2: deleted %d \n", i);
        }
    }
}

int main() {
    //create thread with test functions and print list
    node_t* head = initialize_node(-5);
    node_t* t = initialize_node(-3);
    initialise_list(head, t);
    pthread_t thread1, thread2, thread3, thread4;
    int t1 = pthread_create(&thread1,NULL,test1,NULL);
    printf("t1: %d\n",t1);
    int t2 = pthread_create(&thread2,NULL,test2,NULL);
    printf("t2: %d\n",t2);
    int t3 = pthread_create(&thread3,NULL,test1,NULL);
    printf("t1: %d\n",t3);
    int t4 = pthread_create(&thread4,NULL,test2,NULL);
    printf("t2: %d\n",t4);
    pthread_join(thread1, NULL);
    pthread_join(thread2, NULL);
    pthread_join(thread3, NULL);
    pthread_join(thread4, NULL);
    print_list();
}
```

`userCode.h`
```h
//
// Created by keanukreed on 01.05.23.
//
#include <stdatomic.h>
#include <stdio.h>
#include <stdlib.h>

#ifndef HARRISLIST_HARRISLIST_H
#define HARRISLIST_HARRISLIST_H

typedef struct node{
    _Atomic int key;
    struct node* _Atomic next;
    _Atomic int mark;
    _Atomic int count; //reference counter
}node_t;

typedef struct list{
    node_t* head;
    node_t* tail;
}list_t;

//implementation of reference counter
static inline void
ref_inc(const struct node *ref)
{
    atomic_fetch_add((_Atomic int *)&ref->count, 1); //increases reference counter
}

static inline void
ref_dec(const struct node *ref)
{
    if (atomic_fetch_sub((_Atomic int *)&ref->count, 1) == 1){ //decreases ref counter and frees if 1
        free((void*)ref);
    }

}

//useable List functions
int insert(int key);
int delete(int key);
int find(int key);
void print_list();

//non useable list function
node_t* search(node_t **left, int key);

//Helper functions
int is_marked_node(node_t* node);
void set_unmarked_node(node_t* node);
void set_marked_node(node_t* expected);
int initialise_list(node_t* head, node_t* tail);
node_t* initialize_node(int key);
int mark_neighbour_node(node_t* given, node_t* expected);

#include "userCode.c"

#endif //HARRISLIST_HARRISLIST_H

```

`userCode.c`

```c
#include "userCode.h"
//
// Created by keanukreed on 01.05.23.
//


//testing with only one static list
list_t list;

//fills head and tail in list
int initialise_list(node_t* head, node_t* tail){
    head->next = tail;
    list.head = head;
    list.tail = tail;
    return 1;
}

//creates node and initialises it
node_t* initialize_node(int key){
    node_t* tmp = malloc(sizeof(node_t));
    tmp->key = key;
    tmp->mark = 0;
    tmp->next = NULL;
    tmp->count = 1;
    return tmp;
}

//inserts node based on key and location
int insert(int key){
    //create helper and new node
    node_t *new_node, *right_node, *left_node;
    new_node = initialize_node(key);
    //increase new node ref
    ref_inc(new_node);
    do{
        //pinpoint location for new node
        right_node = search(&left_node, key);
        //check if node with similar key is in list
        if((right_node != list.tail) && (right_node->key == key)){
            return 0;
        }
        //attach right part to new node
        new_node->next = right_node;
        //attach right part of left node to new node
        if(atomic_compare_exchange_strong(&(left_node->next), &right_node, new_node)){
            //list was not modified new node can be inserted
            return 1;
        }
        //list was modified while inserting need to relocate node position
    }while(1);
}

node_t* search(node_t **left, int key){
    //create helper nodes
    node_t *left_next, *right_node;

    search_again:
    do{
        //set temp nodes to iterate through list
        node_t* temp = list.head;
        node_t* temp_next = list.head->next;

        //find left and right node
        do{
            //if node not marked set left bounds
            if(!atomic_load(&(temp_next->mark))){
                (*left) = temp;
                left_next = temp_next;
            }
            //iterate and unmark iterated node
            temp = temp_next;
            atomic_store(&(temp->mark), 0);
            //if iterated through complete list end loop
            if(temp == list.tail){
                break;
            }
            //set temp_next
            temp_next = temp->next;
        }while(atomic_load(&temp_next->mark) || atomic_load(&temp->key) < key);
        //do this until node is either marked or key does not fit anymore, temp is our right bound
        right_node = temp;

        //check if nodes are next to each other
        if(left_next == right_node){
            //check if new node has been logically deleted
            if((right_node != list.tail) && is_marked_node(right_node->next)){
                //list has been modified there is a logically deleted node now
                goto search_again;
            } else {
                //list is in consistent state nodes have been set correctly
                return right_node;
            }
        }

        //error while setting left bound, repeat first step
        if(*left == NULL){
            continue;
        }

        //physically delete logically deleted nodes
        //save current node
        node_t* to_free = left_next;
        //check if marked node is still there
        if(atomic_compare_exchange_strong(&((*left)->next), &left_next, right_node)){
            //delete marked node
            ref_dec(to_free);
            //check if there is another logically deleted node or if list has been modified
            if((right_node != list.tail) && is_marked_node(right_node->next)){
                goto search_again;
            } else {
                return right_node;
            }
        }
    }while(1);
}

int delete(int key){
    //create helper nodes
    node_t *right_node, *right_node_next, *left_node, tmp;

    do{
        //pinpoint location of to be deleted
        right_node = search(&left_node, key);
        //key is not in list
        if((right_node == list.tail) || (right_node->key != key)){
            return 0;
        }
        //check if wanted already logically deleted
        right_node_next = right_node->next;
        if(!is_marked_node(right_node_next)){
            //create marked version of right next, tmp node
            tmp = *right_node_next;
            set_marked_node(&tmp);
            //if list wasn`t modified swap node to its marked version
            if(atomic_compare_exchange_strong(&(right_node->next),&right_node_next,&tmp)){
                break;
            }
        }
    } while(1);
    //if node isnt already phisically deleted delete it through search function
    if(!atomic_compare_exchange_strong(&(left_node->next), &right_node, right_node_next)){
        right_node = search(&left_node, right_node->key);
    }
    //node can be decreased and freed if not used
    ref_dec(right_node);
    return 1;
}

int find(int key){
    //create helper nodes
    node_t *right_node, *left_node;
    //pinpoint wanted node location
    right_node = search(&left_node, key);
    //check if wanted node was found
    if((right_node == list.tail) || (right_node->key != key)) {
        return 0;
    } else {
        return 1;
    }
}


//helper functions (some currently modiefied or unused)
int is_marked_node(node_t* node){
    return atomic_load(&node->mark);
}

void set_unmarked_node(node_t* node){
    atomic_store(&(node->mark), 0);
}

void set_marked_node(node_t* node){
    atomic_store(&(node->mark), 1);
}

int mark_neighbour_node(node_t* given, node_t* expected){
    return NULL;
}

void print_list(){
    node_t* tmp = list.head;
    while(tmp != NULL){
        printf("%d\n",tmp->key);
        tmp = tmp->next;
    }
}
```

Call dartagnan with the command:

```
java -jar <path to dat3m jar> tso.cat --target=tso <path to main.c> --property=program_spec --bound=2 --witness.graphviz=true --modeling.threadCreateAlwaysSucceeds=true
```

The full program output looks like this:

```
[20.06.2023] 15:02:13 [32m[INFO][m GitInfo.CreateGitInfo - Git branch: master
[20.06.2023] 15:02:13 [32m[INFO][m GitInfo.CreateGitInfo - Git commit ID: e59dbcb928030380ef62602879a34e2f62817cd5
[20.06.2023] 15:02:13 [32m[INFO][m GlobalSettings.LogGlobalSettings - ARCH_PRECISION: -1
[20.06.2023] 15:02:13 [32m[INFO][m GlobalSettings.LogGlobalSettings - ATOMIC_AS_LOCK: false
[20.06.2023] 15:02:13 [32m[INFO][m GlobalSettings.LogGlobalSettings - REFINEMENT_GENERATE_GRAPHVIZ_DEBUG_FILES: false
[20.06.2023] 15:02:13 [32m[INFO][m GlobalSettings.LogGlobalSettings - REFINEMENT_SYMMETRY_LEARNING: FULL
[20.06.2023] 15:02:13 [32m[INFO][m Dartagnan.main - Program path: /home/m_jung/SEP/sourcecode/output/test-runs/6a5065bd-e192-4760-83b8-0449b11d4737/TCS1/main.c
[20.06.2023] 15:02:13 [32m[INFO][m Dartagnan.main - CAT file path: /home/m_jung/dartagnan/Dat3M/cat/tso.cat
[20.06.2023] 15:02:15 [32m[INFO][m VisitorBoogie.visitLine_comment - SMACK version: 2.8.0 
[20.06.2023] 15:02:15 [32m[INFO][m VisitorBoogie.visitMain - Number of threads (including main): 5
[20.06.2023] 15:02:15 [32m[INFO][m RefinementSolver.run - refinement.baseline: []
[20.06.2023] 15:02:15 [32m[INFO][m UnreachableCodeElimination.run - #Events before DCE: 3523
[20.06.2023] 15:02:15 [32m[INFO][m UnreachableCodeElimination.run - #Events after DCE: 3332
[20.06.2023] 15:02:15 [32m[INFO][m ComplexBlockSplitting.run - Split 181 complex blocks.
[20.06.2023] 15:02:15 [32m[INFO][m BranchReordering.run - Branches reordered
[20.06.2023] 15:02:15 [32m[INFO][m BranchReordering.run - program.processing.detReordering: true
[20.06.2023] 15:02:15 [1;31m[ERROR][m Dartagnan.main - Loop entry label 12:$bb3: has multiple back edges
com.dat3m.dartagnan.exception.MalformedProgramException: Loop entry label 12:$bb3: has multiple back edges
	at com.dat3m.dartagnan.program.processing.LoopFormVerification.checkAndCountLoops(LoopFormVerification.java:62) ~[dartagnan-3.1.1.jar:?]
	at java.util.stream.ReferencePipeline$4$1.accept(ReferencePipeline.java:212) ~[?:?]
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.IntPipeline.reduce(IntPipeline.java:491) ~[?:?]
	at java.util.stream.IntPipeline.sum(IntPipeline.java:449) ~[?:?]
	at com.dat3m.dartagnan.program.processing.LoopFormVerification.run(LoopFormVerification.java:41) ~[dartagnan-3.1.1.jar:?]
	at com.dat3m.dartagnan.program.processing.ProcessingManager.lambda$run$0(ProcessingManager.java:110) ~[dartagnan-3.1.1.jar:?]
	at java.util.ArrayList.forEach(ArrayList.java:1541) ~[?:?]
	at com.dat3m.dartagnan.program.processing.ProcessingManager.run(ProcessingManager.java:110) ~[dartagnan-3.1.1.jar:?]
	at com.dat3m.dartagnan.verification.solving.ModelChecker.preprocessProgram(ModelChecker.java:76) ~[dartagnan-3.1.1.jar:?]
	at com.dat3m.dartagnan.verification.solving.RefinementSolver.run(RefinementSolver.java:104) ~[dartagnan-3.1.1.jar:?]
	at com.dat3m.dartagnan.verification.solving.RefinementSolver.run(RefinementSolver.java:90) ~[dartagnan-3.1.1.jar:?]
	at com.dat3m.dartagnan.Dartagnan.main(Dartagnan.java:172) [dartagnan-3.1.1.jar:?]
```

**Comment by ThomasHaas** [comment]
The issue is that the loops do not get normalized by LLVM.
You can add the option `-loop-simplify` into `ATOMIC_REPLACE_OPTS` to force LLVM to normalize the loops.
```
export ATOMIC_REPLACE_OPTS="-mem2reg -sroa -early-cse -indvars -loop-unroll -simplifycfg -gvn -loop-simplify"
```
That being said, this benchmark is far out-of-reach for Dartagnan even with just `Bound=2`. The resulting program has over 13000 events and just generating the encoding might take hours (if it even finishes).

**Comment by hernanponcedeleon** [comment]
Fixed by 907e0110d228dcee705dda362925f04893eff5e9
